src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,363 @@
+
+/**
+ * Shuffle the data.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::Shuffle()
+{
+  // Determine new ordering.
+  arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
+      dataset.n_cols - 1, dataset.n_cols));
+
+  // Re-sort data.
+  arma::mat newData = dataset.cols(ordering);
+  math::ClearAlias(dataset);
+  dataset = std::move(newData);
+
+  // Assemble data for batch constructor.  We need reverse orderings though...
+  arma::uvec reverseOrdering(ordering.n_elem);
+  for (size_t i = 0; i < ordering.n_elem; ++i)
+    reverseOrdering[ordering[i]] = i;
+
+  arma::umat newLocations(2, groundTruth.n_nonzero);
+  arma::vec values(groundTruth.n_nonzero);
+  arma::sp_mat::const_iterator it = groundTruth.begin();
+  size_t loc = 0;
+  while (it != groundTruth.end())
+  {
+    newLocations(0, loc) = reverseOrdering(it.col());
+    newLocations(1, loc) = it.row();
+    values(loc) = (*it);
+
+    ++it;
+    ++loc;
+  }
+
+  groundTruth = arma::sp_mat(newLocations, values, groundTruth.n_rows,
+                             groundTruth.n_cols);
+}
+
+template <typename MatType>
+double LinearSVMFunction<MatType>::Evaluate(
+    const arma::mat& parameters)
+{
+  // The objective function is the hinge loss function and it is
+  // calculated over all the training examples.
+
+  // Calculate the loss and regularization terms.
+  // L_i = Σ_i Σ_m max(0, Δ + (w_m x_i + b_m) - (w_{y_i} x_i + b_{y_i}))
+  // where (m != y_i)
+  double loss, regularization, cost;
+
+  // Scores for each class are evaluated.
+  arma::mat scores = parameters * dataset;
+
+  // Make a matrix where each column holds the value of the
+  // correct class Score.
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+
+  // Evaluates the margin and makes (y_i) term 0 so that
+  // they don't have any contribution in the summation.
+  arma::mat margin = scores - correctScores + delta -
+      (delta * groundTruth);
+
+  // The Hinge Loss Function
+  loss = arma::accu(arma::clamp(margin, 0.0, DBL_MAX));
+  loss /= dataset.n_cols;
+
+  // Adding the regularization term.
+  regularization = 0.5 * lambda * arma::dot(parameters,
+      parameters);
+
+  cost = loss + regularization;
+  return cost;
+}
+
+template <typename MatType>
+double LinearSVMFunction<MatType>::Evaluate(
+    const arma::mat& parameters,
+    const size_t firstId,
+    const size_t batchSize)
+{
+  const size_t lastId = firstId + batchSize - 1;
+
+  // Calculate the loss and regularization terms.
+  double loss, regularization, cost;
+
+  arma::mat scores = parameters * dataset;
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+  arma::mat margin = scores - correctScores + delta
+      - (delta * groundTruth);
+
+  // The Hinge Loss Function
+  loss = arma::accu(arma::clamp(margin.cols(firstId, lastId),
+      0.0, DBL_MAX));
+  loss /= batchSize;
```

Comment:
Using the columns `firstId` through `lastId` was having several issues,but I cannot recall those. However, I'll go through this again.

Reply:
Ok, I just recalled the issue. On using `cols()` on a `arma::sp_mat`, everything turns to 0. 

I had tried to use the way you suggested way back but had to do the current implementation because of this.

```
  std::cout << "Before:\n" << arma::conv_to<arma::mat>::from(groundTruth);
  std::cout << "After:\n" << arma::conv_to<arma::mat>::from(groundTruth.cols(0,2));
```

```
Before:
        0   1.0000        0
   1.0000        0   1.0000
After:
        0        0        0
        0        0        0
```

There is no issue with `conv_to`. I've checked that.

Reply:
I can't reproduce this.  I'm using this test program:

```
#include <armadillo>

using namespace arma;

int main()
{
  arma::sp_mat x(2, 3);
  x(1, 0) = 1;
  x(0, 1) = 2;
  x(1, 2) = 3;

  std::cout << "Sparse form:\n";
  x.print();

  std::cout << "\nConstructed into a dense matrix:\n";
  arma::mat z(x.cols(0, 2));
  z.print();

  std::cout << "\nconv_to<arma::mat>::from():\n";
  std::cout << arma::conv_to<arma::mat>::from(x.cols(0, 2));
}
```

That gives 

```
Sparse form:
[matrix size: 2x3; n_nonzero: 3; density: 50.00%]

     (1, 0)         1.0000
     (0, 1)         2.0000
     (1, 2)         3.0000


Constructed into a dense matrix:
        0   2.0000        0
   1.0000        0   3.0000

conv_to<arma::mat>::from():
        0   2.0000        0
   1.0000        0   3.0000
```

What happens if you build and run that test program?

Reply:
This is strange, I'll try to run it myself.

Reply:
The code that you sent me gave the same result as yours.

But when I run the same code for `groundTruth` it gives `0` for every value.
```
#include <iostream>
#include <armadillo>

void GetGroundTruthMatrix(
        const arma::Row<size_t>& labels,
        arma::sp_mat& groundTruth,
        size_t numClasses)
{
  arma::uvec rowPointers(labels.n_elem);
  arma::uvec colPointers(labels.n_elem + 1);

  for (size_t i = 0; i < labels.n_elem; i++)
  {
    rowPointers(i) = labels(i);
    colPointers(i + 1) = i + 1;
  }

  arma::vec values;
  values.ones(labels.n_elem);

  groundTruth = arma::sp_mat(rowPointers, colPointers, values, numClasses,
                             labels.n_elem);
}


int main()
{
  arma::Row<size_t> labels = "1 0 1";

  arma::sp_mat x;
  GetGroundTruthMatrix(labels, x, 2);

  std::cout << "Sparse form:\n";
  x.print();

  std::cout << "\nConstructed into a dense matrix:\n";
  arma::mat z(x.cols(0, 2));
  z.print();

  std::cout << "\nconv_to<arma::mat>::from():\n";
  std::cout << arma::conv_to<arma::mat>::from(x.cols(0, 2));

  return 0;
}
```
This gives
```
Sparse form:
[matrix size: 2x3; n_nonzero: 3; density: 50.00%]

     (1, 0)         1.0000
     (0, 1)         1.0000
     (1, 2)         1.0000


Constructed into a dense matrix:
        0        0        0
        0        0        0

conv_to<arma::mat>::from():
        0        0        0
        0        0        0
```

Reply:
Could you test this one out?

Reply:
What version of Armadillo are you using?  When I run that exactly as you gave it, I see

```
Using GroundTruth values

Sparse form:
[matrix size: 2x3; n_nonzero: 3; density: 50.00%]

     (1, 0)         1.0000
     (0, 1)         1.0000
     (1, 2)         1.0000


Constructed into a dense matrix:
        0   1.0000        0
   1.0000        0   1.0000

conv_to<arma::mat>::from():
        0   1.0000        0
   1.0000        0   1.0000


Using Custom values
Sparse form:
[matrix size: 2x3; n_nonzero: 3; density: 50.00%]

     (1, 0)         1.0000
     (0, 1)         2.0000
     (1, 2)         3.0000


Constructed into a dense matrix:
        0   2.0000        0
   1.0000        0   3.0000

conv_to<arma::mat>::from():
        0   2.0000        0
   1.0000        0   3.0000
```

I compiled as `g++ -o arma_test arma_test.cpp -larmadillo`, with four different versions of Armadillo: 8.500.1, 9.200.7, 7.400.3, and 6.700.6, and none of those gave the wrong output for converting to a dense matrix.  They all gave the output above.  I'm using gcc 8.2.0.  I even tried with clang and can't reproduce it.

Reply:
Uhh, I'm using Armadillo: 9.100.5 and gcc 8.2.0. 
Thanks for checking this out on so many versions of Armadillo and looking into it, I'll look for the issue since it is clear that there is some issue with something on my side.

Reply:
I might not include any changes to this as of now. I am still having this issue and I'm not able to test this out( the test fails because all the `sp_mat` where I use gets converted to `0`). I've tried reinstalling several versions of Armadillo and this is super strange imo.

Reply:
We cannot merge this until this issue is fixed.  This will be orders of magnitude slowdown for SGD-like optimizers.  So we need to figure out what is happening on your system that is different.

Reply:
I'll try to work it out.

Reply:
Maybe do you want to try in a VirtualBox Linux container or something like this?  Actually, are you on OS X or Linux?  If the OS is different I suppose that might be the issue.  Still extremely strange, I haven't managed to think of anything that could possibly cause this.

Reply:
@rcurtin I probably looked alot about this issue. Here is everything I have done in details. 
I have a system with Ubuntu 18.04 on it. I have tried using Armadillo version `9.100.5` and version `8.400`. I was wondering if there were some issues with me using `OpenBLAS` or `LAPACK` so I tried them both and couldn't get a proper result (I may have done something wrong here, so it could also be the case). At this point, I pretty much messed up a lot with those libraries and dependencies, and I could not debug anything (I was also confused at this point). So I did a factory reset, installed a fresh copy of Ubuntu and tried it again. As soon as my system was ready, I got the Armadillo from the package manager `$ sudo apt-get install libarmadillo-dev` (so this time, the dependencies for Armadillo were installed by the package manager). At this point the version of Armadillo I have is `8.400`. On running the above program, I did not get the desired results.
After this, I tried to install `Armadillo` on my friend's system which runs on `Linux Arch`. Here, I got the package from the package manager (https://aur.archlinux.org/packages/armadillo/) and all the dependencies were managed by the package manager. Then, I ran the above program and got the same problem.

After all this, I made some changes **only** to separable `Evaluate()`, and checked them by committing them one after the other (by checking the result of the Travis Build).
  - Test Case 1:
    - In the first commit, I made the changes **(only to separable `Evaluate()`)** that were required, like using `dataset.cols(firstId, lastId)` and similar changes to get it working.
    - One important change in this was `arma::conv_to<arma::mat>::from(groundTruth).cols(firstId, lastId)`, which was basically giving me what `groundTruth.cols(firstId, lastId)` was supposed to give me, but which it wasn't.
    - I also made a test case (LinearSVMTestCase1) where I test separable `Evaluate()` and the overall `Evaluate()` on a very small dataset which can be evaluated by hand.
    - Result:
      - Commit: 7613bf9
      - It passes all the tests on my system.
      - Travis build is successful with no errors in test (https://travis-ci.org/mlpack/mlpack/builds/506116717)

  - Test Case 2:
    - Keeping everything else same, I replaced `arma::conv_to<arma::mat>::from(groundTruth).cols(firstId, lastId)` with `groundTruth.cols(firstId, lastId)`.
    - Result:
      - Commit: 828b07a
      - Test `LinearSVMTestCase1` and `LinearSVMFunctionSeparableEvaluate` fails on my system, since they use that `Evaluate()`.
      - These tests also fail in the Travis Build (https://travis-ci.org/mlpack/mlpack/jobs/506178863). Also, they fail with the same values of `hingeLoss` (comparing that of my system and the one showing in logs).

I'm not sure if it was the right way to reproduce this error, but since I got this issue on 2 system 
 (my freshly reinstalled `Ubuntu 18.04` and `Arch Linux`), which had no `Armadillo` installed before me testing it out (and also, got `Armadillo` from the Package Manager in these cases). 

If I did anything wrong here, or if I did any mistake here, please let me know.

Reply:
@rcurtin could you take a look at this?

Reply:
Yes, it's on my todo list, I haven't had a chance yet.

Reply:
Ah, alright, I thought that you missed my last comment there, so I just pinged you again. 
Thanks for letting me know :)